blog/2020-10-09-hs-to-coq/2020-10-09-hs-to-coq.md

```diff
@@ -0,0 +1,1172 @@
+---
+title: "Tutorial: Verify Haskell Programs with hs-to-coq"
+author: Li-yao Xia
+description: A tutorial on hs-to-coq
+tags: haskell, coq
+withtoc: true
+alectryon: [-Q, "blog/2020-10-09-hs-to-coq/lib/Src", Src, -Q, "blog/2020-10-09-hs-to-coq/lib/Proofs", Proofs]
+alectryon-cache: "blog/2020-10-09-hs-to-coq/alectryon/"
+---
+
+## Preface
+
+Programming is the power of automation at our fingertips.
+A task, a program, can be run over and over, billions of instructions
+per second. And with great power comes great responsibility.
+A program can also fail, over and over, billions of instructions per second.
+Many solutions have been devised to reduce the likelihood and impact of human
+error in programming, testing and type systems being prominent examples.
+Programming is also a domain where pure logic prospers especially well.
+Programs are mathematical objects that we can prove properties about,
+and unlike other areas of expertise, the gap between our mental model and the
+program that actually runs can be vanishingly small.
+
+Proving programs correct is possible, but it is also very hard.
+A rigorous proof requires knowledge and discipline, whose acquisition takes
```

Comment:
```suggestion
Writing a rigorous proof requires knowledge and discipline which take years of expensive training and experience to acquire. It requires sustained effort and perseverance to succeed.
```
or something like that. Your sentence structure here just loses the subject, I think.